### PR TITLE
Get prefers color scheme from accountsservice

### LINKED ISSUE #276: Add way to check if the OS prefers a dark style
To enable system apps/components to more easily go dark when the OS wants to go dark. See: elementary/switchboard-plug-pantheon-shell#171

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/71265927-add-way-to-check-if-the-os-prefers-a-dark-style?utm_campaign=plugin&utm_content=tracker%2F60830613&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60830613&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by danrabbit**
Create a Granite.Settings class like Gtk.Settings. Fixes #276 

Read-only is intentional as we don't want apps setting this for the user

### CODE REVIEW

File: lib/Widgets/Settings.vala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        public abstract string prefers_color_scheme { owned get; set; }
+        public abstract string time_format { owned get; set; }
+    }
+
+    [DBus (name = "org.freedesktop.Accounts")]
+    interface FDO.Accounts : Object {
+        public abstract string find_user_by_name (string username) throws GLib.Error;
+    }
+
+    public class Settings : Object{
```
Comment (cassidyjames) [inline lib/Widgets/Settings.vala]: Lint

```suggestion
    public class Settings : Object {
```

**Comment by cassidyjames** [comment]
We'll want to switch the AccountsService side to an enum since https://github.com/elementary/default-settings/pull/166 was merged.

**Comment by danrabbit** [comment]
@cassidyjames Updated. This should work with the new accountsservice stuff.

Since there's no system UI for this yet, you can send this in dfeet `"io.elementary.pantheon.AccountsService", "PrefersColorScheme", GLib.Variant("u", 1)`

I guess there's a question of how should we demo this? I think we want the override here for the demo maybe? Or should we remove that toggle?

Also, we currently only include this for accountsservice in Odin. Should we add this in Hera too? Maybe not necessarily exposing the setting yet, but if it exists in Granite and AccountsService we could start testing it

Is it okay to make this a boolean property or should we make this enum out of the box? The boolean is nice because you can super easily bind this to app-prefers-dark right now. But this doesn't account for any prefers-light. I guess if we decide to support that we can come back and deprecate the boolean property?

**Comment by cassidyjames** [comment]
>I guess there's a question of how should we demo this? I think we want the override here for the demo maybe? Or should we remove that toggle?

I think we default to the user preference but still allow a local override (so devs can still test how widgets look without changing their whole system preference)? But we can think about it. Another option would be to implement what we expect to see in apps like Code or Terminal where we might want to allow for a tri-state of light/dark/follow system.

>Also, we currently only include this for accountsservice in Odin. Should we add this in Hera too? Maybe not necessarily exposing the setting yet, but if it exists in Granite and AccountsService we could start testing it

Yeah I think we include it in Hera. I think I'd like to ship the system UI stuff for 5.2 if possible, because it has been a long time coming and would be a nice preview of the bigger dark style stuff for 6.0.

>Is it okay to make this a boolean property or should we make this enum out of the box? The boolean is nice because you can super easily bind this to app-prefers-dark right now. But this doesn't account for any prefers-light. I guess if we decide to support that we can come back and deprecate the boolean property?

Yeah I think boolean is fine and we can deprecate later if/when we expose a light preference.

**Comment by cassidyjames** [comment]
Hmm, I guess this raises the question of how apps should bind to it. If it changes live, I imagine we'd want the app to change instantly, too. But if they offer an override, how does that interact? I think that's where it has to become a tri-state of:

- Follow system preference
- Always light
- Always dark

Which is similar to what I see in Android apps fwiw. I think _most_ apps that support a dark style should just automatically follow along with the setting, though.

**Comment by cassidyjames** [comment]
I guess I'm fine in general with this as a starting point, we can always iterate. I would like to get a review of the code side from someone more familiar and opinionated about Granite and namespacing and everything, though.

File: lib/Widgets/Settings.vala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    "org.freedesktop.Accounts",
+                    user_path,
+                    GLib.DBusProxyFlags.GET_INVALIDATED_PROPERTIES
+                );
+
+                user_prefers_dark = pantheon_act.prefers_color_scheme == 1;
+
+                ((GLib.DBusProxy) pantheon_act).g_properties_changed.connect ((changed_properties, invalidated_properties) => {
+                    uint prefers_color_scheme;
+                    changed_properties.lookup ("PrefersColorScheme", "u", out prefers_color_scheme);
```
Comment (tintou) [inline lib/Widgets/Settings.vala:86]:
```suggestion
                    int prefers_color_scheme;
                    changed_properties.lookup ("PrefersColorScheme", "i", out prefers_color_scheme);
```

**Comment by tintou** [comment]
It would probably make it clearer indeed to return an enum instead of having a boolean, that way we can extend it without fearing API breaks

**Comment by cassidyjames** [comment]
@tintou see: https://github.com/elementary/default-settings/pull/167#discussion_r414022506

We should decide on unsigned or signed and then make sure it matches in the master and hera branches of default-settings as well as here.

**Pull request merged**